scripts/train_causal.py

```diff
@@ -0,0 +1,312 @@
+    type=str,
+    default=None,
+    help="Start/resume from a checkpoint a the given path",
+)
+parser.add_argument(
+    "--tokenizer",
+    type=str,
+    default="char_tokenizer",
+    help="Name of or path to the tokenizer (e.g. ~/tokenizer.model)",
+)
+
+parser.add_argument(
+    "--peft_method",
+    type=str,
+    default=None,
+    help="Peft method (lora, ...). Default None if not using peft",
+)
+
+parser.add_argument(
+    "--dataset_style",
+    type=str,
+    default="instruction",
+    help="'instruction' uses alpaca-formatted json. 'text' points to a raw text file.",
+)
+parser.add_argument(
+    "--dataset_path",
+    type=str,
+    default="https://raw.githubusercontent.com/tatsu-lab/stanford_alpaca/main/alpaca_data.json",
+    help="The path or URI refering to data to use in tuning or training",
+)
+
+
+args = parser.parse_args()
+
+local_rank = int(os.getenv("LOCAL_RANK", 0))
+world_size = int(os.getenv("WORLD_SIZE", 1))
+
+# default search for what's available
+device_type = args.device_type
+if device_type is None:
+    if torch.cuda.is_available():
+        device_type = "cuda"
+    elif torch.backends.mps.is_available():
+        device_type = "mps"
+    else:
+        device_type = "cpu"
+
+device = torch.device(device_type, local_rank)
+
+group = None
+
+if args.distributed is not None:
+    dist.init_process_group(backend="nccl", rank=local_rank, world_size=world_size)
+    group = dist.GroupMember.WORLD
+
+
+def get_loss_fn():
+    ce_loss = torch.nn.CrossEntropyLoss()
+
+    def loss_fn(output, target):
+        # model output is batch x seq_len x vocab_size.
+        # ce expects batch x vocab x seq_len
+        output = output.transpose(-1, -2)
+        return ce_loss(output, target)
+
+    return loss_fn
+
+
+def _hf_to_fms_model(model):
+    model.orig_forward_f = model.forward
+
+    def new_forward(self, *args, **kwargs):
+        hf_out = model.orig_forward_f(self, *args, **kwargs)
+        logits = hf_out.logits
+        kv_cache = hf_out.past_key_values
+        if "use_cache" in kwargs and kwargs["use_cache"]:
+            return logits, kv_cache
+        else:
+            return logits
+
+    model.forward = new_forward
+    return model
+
+
+def peft_model(model):
+    # TODO: consider using loralib directly instead:
+    # https://github.com/microsoft/LoRA/tree/main
+    # This would simplify checkpoint handling.
+    """
+    Converts an fms model to an PEFT and HF adapted (wrapped) model, while
+    preserving the original `forward` function.
+
+    If we call state_dict() on one of these models we'll get keys prefixed with
+    `base_model.model.*`, so correctly saving and re-loading one of these
+    requires some care. The state_dict will also contain all paramters, not
+    just the adapter, though we plan to use merged tuned models for now.
+    """
+    from fms.models.hf.llama.modeling_llama_hf import HFAdaptedLLaMAForCausalLM
+
+    model = HFAdaptedLLaMAForCausalLM.from_fms_model(model)
```

Comment:
we could use `to_hf_api` here that way we don't need to know the underlying model

Reply:
done